File: DigitalImageProcessing/python/polar_transform_img.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 利用极坐标变换对图像进行变换，校正图像中的圆形区域
OpenCV 实现了线性极坐标变换和对数极坐标变换。
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-09-15
"""

import sys
import numpy as np
import cv2 as cv


def polar(input_img, center, r, theta=(0, 360), r_step=1.0, theta_step=360.0/(180*8)):
    """实现图像的极坐标变换。

    Args:
        input_img (ndarray): 输入图像
        center ([type]): 极坐标变换中心
        r (二元元组 tuple): 最小距离和最大距离
        theta (tuple, optional): 角度范围. Defaults to (0, 360).
        r_step (float, optional): r 极长进行离散化的步长. Defaults to 1.0.
        theta_step ([type], optional): theta 角度进行离散化的步长. Defaults to 360.0/(180*8).

    Returns:
        [ndarray]: 输出图像
    """
    min_r, max_r = r
    min_theta, max_theta = theta
    cx, cy = center
    # 输出图像的 H W
    H = int((max_r - min_r) / r_step) + 1
    W = int((max_theta - min_theta) / theta_step) + 1
    output_img = 125 * np.ones((H, W), input_img.dtype)

    # 极坐标变换
    r = np.linspace(min_r, max_r, H)
    r = np.tile(r, (W, 1))
    r = np.transpose(r)
    theta = np.linspace(min_theta, max_theta, W)
    theta = np.tile(theta, (H, 1))
    x, y = cv.polarToCart(r, theta, angleInDegrees=True)

    # 插值算法 最邻近插值
    for i in range(H):
        for j in range(W):
            px = int(round(x[i][j] + cx))
            py = int(round(y[i][j] + cy))
            if (px>=0 and px<=W-1) and (py>=0 and py<=H-1):
                output_img[i][j] = input_img[i][j] 

    return output_img


# --------------------------
if __name__ == "__main__":
    # 0. tips for coding with Numpy
    # The function tile() of numpy.
    a = np.array([[1, 2], [3, 4]])
    # 将 a 分别再垂直方向和水平方向复制 2 次和 3 次
    b = np.tile(a, (2, 3))
    print(f"The function tile of Numpy: \n{b}")

    # 1. 极坐标变换对图像
    if len(sys.argv) > 1:
        input_img = cv.imread(sys.argv[1], 0)
        h, w = input_img.shape[:2]
        cx, cy = 508, 503
        cv.circle(input_img, (int(cx), int(cy)), 10, (255.0, 0, 0), 3)
        output_img = polar(input_img=input_img, center=(cx, cy), r=(200, 550))
        # 翻转
        output_img = cv.flip(output_img, 0)

        cv.imshow("OriginImage", input_img)
        cv.imshow("OutputImage", output_img)
        cv.waitKey(0)
        cv.destroyAllWindows()

    else:
        print(f"Usage: python polar image.")

File: DigitalImageProcessing/cpp/polar_transform_img.cpp
/**
 * @File    : polar_transform_img.cpp
 * @Brief   : 利用极坐标变换对图像进行变换，校正图像中的圆形区域
 * OpenCV 实现了线性极坐标变换和对数极坐标变换。
 * @Author  : Wei Li
 * @Date    : 2021-09-15
*/

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

cv::Mat polar(cv::Mat input_img, cv::Point2f center, cv::Size size, float minR=0, float minTheta=0, float thetaStep=1.0/4, float rStep=1.0)
{
    // 1. 构建 r
    cv::Mat ri = cv::Mat::zeros(cv::Size(1, size.height), CV_32FC1);
    for (int i = 0; i < size.height; ++i)
    {
        ri.at<float>(i, 0) = minR + i * rStep;
    }
    cv::Mat r = cv::repeat(ri, 1, size.width);

    // 2. 构建 theta
    cv::Mat thetaj = cv::Mat::zeros(cv::Size(size.width, 1), CV_32FC1);
    for (int j = 0; j < size.width; ++j)
    {
        thetaj.at<float>(0, j) = minTheta + j*thetaStep;
    }
    cv::Mat theta = cv::repeat(thetaj, size.height, 1);

    // 3. 将极坐标转换为笛卡尔坐标
    cv::Mat x, y;
    cv::polarToCart(r, theta, x, y, true);
    // 将坐标原点移动到中心点
    x += center.x;
    y += center.y;
    // 最邻近插值
    cv::Mat dst = 125 * cv::Mat::ones(size, CV_8UC1);
    for (int i = 0; i < size.height; ++i)
    {
        for (int j = 0; j < size.width; ++j)
        {
            float xij = x.at<float>(i, j);
            float yij = y.at<float>(i, j);
            int nearestx = int(std::round(xij));
            int nearesty = int(std::round(yij));
            if ((0 <= nearestx && nearestx < input_img.cols) && (0 <= nearesty && nearesty < input_img.rows))
            {
                dst.at<uchar>(i, j) = input_img.at<uchar>(nearestx, nearesty);
            }
        }
    }
    return dst;
}

// -------------------------------
int main(int argc, char** argv)
{
    // tips for coding, 实现矩阵的平铺 repeat(tile() function in Numpy)
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }

        // 图像的极坐标变换
        float thetaStep = 1.0 / 4;
        float minR = 270;
        cv::Size size(int(360 / thetaStep), 70);
        cv::Mat dst = polar(image, cv::Point2f(508, 503), size, minR);
        
        // 沿着水平方向的镜像处理，翻转
        cv::flip(dst, dst, 0);
        cv::imshow("OriginImage", image);
        cv::imshow("PolarImage", dst);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV warpAffine image." << std::endl;
        return -1;
    }
}
